### ISSUE TITLE
QueryParser fails to accommodate case-sensitive analyzers

### ISSUE THREAD

**Issue by robfrank** [orientechnologies/orientdb#5197]
_From @retrography on April 24, 2015 23:3_

Lucene's QueryParser applies a lower case filter on expanded queries by default (wildcard, fuzzy, etc...), and there doesn't seem to be any configuration in OrientDB to modify this behavior. This prevents the use of all case-sensitive analyzers like KeywordAnalyzer.

It is possible to set this behavior automatically for the well-known case-sensitive analyzers, but given that we can load custom analyzers in OrientDB, there must be a way to configure it manually as well.

Here is the potential fix:

``` Java
queryParser.setLowercaseExpandedTerms(false);
```

You can read about the issue that I have confronted [here](http://stackoverflow.com/questions/29853405) if you would like to reproduce the issue.

_Copied from original issue: orientechnologies/orientdb-lucene#47_

**Comment by robfrank** [comment]
_From @retrography on May 8, 2015 22:54_

The problem persists in 2.1-rc2

**Comment by robfrank** [comment]
_From @retrography on May 28, 2015 22:45_

@maggiolo00 @lvca Given that OrientDB understands JSON documents natively, and that Lucene indexes have way too many options and properties, maybe it is easier to adopt a well known Lucene JSON query syntax like that of elastic search: https://github.com/elastic/elasticsearch/blob/master/rest-api-spec/api/search.json . And the same can be done for advanced Lucene index creation...

The good thing is that elastic search is license-compatibe with OrientDB (Apache2).

I am not sure how much of a help this is, just throwing the idea out there...

**Comment by robfrank** [comment]
_From @retrography on May 29, 2015 14:51_

@maggiolo00 @lvca I patched and recompiled `orientdb-lucene` yesterday with the above mentioned modification and it did not work. The issue with case-sensitive queries persist, but it seems that the suggested solution (setting `LowercaseExpandedTerms`) solves the problem only partially. A simple search in online forums reveals that many people have this issue with Lucene's parsed queries, and it seems they have all solved it using query builders (like `RangeQuery` and `FuzzyQuery`) or `PhraseQuery` instead of `ParsedQuery` and `MultiFieldQueryParser`. I will keep you posted on that if I learn anything new.

**Comment by robfrank** [comment]
_From @maggiolo00 on May 29, 2015 14:54_

Thanks retrography



2015-05-29 16:51 GMT+02:00 retrography <notifications@github.com>:

**Comment by robfrank** [comment]
_From @retrography on May 30, 2015 9:13_

@maggiolo00 @lvca I fixed it, and additionally I added support for for all option setter methods in `QueryParserBase` (https://lucene.apache.org/core/4_7_0/queryparser/org/apache/lucene/queryparser/classic/QueryParserBase.html). The issue is that I can't seem to figure out how to pass a context variable from the SQL query to `OLuceneTextOperator` similar to the way `$spatial` is passed to `OLuceneNearOperator`. If you explain that to me, I will make a pull request tomorrow so the patch can be incorporated in the next release. 

I have tried the following query:

```sql
select * from _tempt where [hash, $parser] lucene ["wemfOGxqCfOTPi0ND~0.8", {"LowercaseExpandedTerms": "false"}]
```

And then the following code in `OLuceneTextOperator` to  grab the value of `$parser`:

```java
Object parser = iContext.getVariable("parser");
Map<String, Object> parserOptions;

if (parser instanceof Map) {
    parserOptions = (Map<String, Object>) parser;
} else {
    parserOptions = new HashMap<String, Object>();
}
```

But this doesn't seem to work. The rest of the patch works.

**Comment by robfrank** [comment]
_From @maggiolo00 on May 30, 2015 11:30_

@retrography 

i will do another things. 

i would like to support this syntax

```
select * from _tempt where [hash] lucene { "q" : "wemfOGxqCfOTPi0ND~0.8", "options"  :{"LowercaseExpandedTerms": "false"} 
```

If you go to branch develop in method 
createFullQuery of OLuceneIndexType

you can parse there additional options and set them to the QueryParser.

**Comment by robfrank** [comment]
_From @retrography on May 31, 2015 4:46_

I just made a pull request. 

@maggiolo00 : I noticed that I have changed the name of `createFullQuery` to `createFullTextQuery` by mistake and that the new syntax does not correspond exactly with what you wanted (no englobing "options" element). Don't hesitate to go ahead and make the proper modifications. But I will also have a bit of time next weekend, and I can fix these myself (if you prefer). 

Sincerely, I am not sure if it is a good idea to envelope the options in the "options" element. It is for sure more document/json-like but it is also for sure very much less sql-like and pretty boring to write down in each query. 

Apart from OLuceneIndexType, I have added a couple of methods to OFullTextCompositeKey as well. I found it the best place to do the parameter processing. I hope that is OK with the overall design you have in mind.

I hope this can make it to 2.1. I am betting on it for my work.
